crl/validator.go

```diff
@@ -20,346 +20,420 @@ package crl
+	v.crlChan <- func() {
+		defer close(result)
+		cp := make(map[string]*revocationList, len(v.crls))
+		for ep, crl := range v.crls {
+			cp[ep] = crl
 		}
+		result <- cp
 	}
+	return <-result
+}
 
-	return true
+// setCRL updates crls in a save way
+func (v *validator) getCRL(endpoint string) (*revocationList, bool) {
+	type resultStruct struct {
+		rl     *revocationList
+		exists bool
+	}
+	result := make(chan resultStruct, 1) // buffer size of 1 so the loop never blocks
+	v.crlChan <- func() {
+		defer close(result)
+		out := resultStruct{}
+		out.rl, out.exists = v.crls[endpoint]
+		result <- out
+	}
+	out := <-result
+	return out.rl, out.exists // make blocking call
 }
 
-// anyCertificateActive returns true if any of the certificates is not expired. If the list is empty, it returns false.
-func anyCertificateActive(certs []*x509.Certificate) bool {
-	result := false
-	for _, cert := range certs {
-		if nowFunc().After(cert.NotAfter) {
-			log.Logger().Warnf("Trust store contains expired certificate: %s (s/n: %s)", cert.Subject.String(), cert.SerialNumber.String())
-		} else {
-			// Certificate is active
-			result = true
-		}
+// setCRL updates crls in a save way
+func (v *validator) setCRL(endpoint string, crl *revocationList) {
+	result := make(chan struct{}, 1) // buffer size of 1 so the loop never blocks
+	v.crlChan <- func() {
+		defer close(result)
+		v.crls[endpoint] = crl
+		result <- struct{}{}
 	}
-	return result
+	<-result // make blocking call
 }
 
-func (v *validator) appendCertificates(certificates []*x509.Certificate) {
-	v.certificatesLock.Lock()
-	defer v.certificatesLock.Unlock()
+// sync tries to update all crls
+func (v *validator) sync() {
+	// get a copy of the CRL map, the contents should not be changed directly.
+	crlMap := v.getCRLs()
 
-	for _, certificate := range certificates {
-		subject := certificate.Subject.String()
+	// TODO: does this need a WaitGroup?
+	wg := &sync.WaitGroup{}
+	wg.Add(len(crlMap))
 
-		if _, ok := v.certificates[subject]; !ok {
-			v.certificates[subject] = certificate
+	// Update in parallel if at least one of the issuers is still valid
+	for endpoint, current := range crlMap {
+		if len(current.issuers) > 0 && !anyCertificateActive(current.issuers...) {
+			wg.Done()
+			continue
 		}
-	}
-}
-
-func (v *validator) VerifyPeerCertificateFunction(maxValidityDays int) func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
-	return func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
-		// Import unknown certificates
-		var verifiedCerts []*x509.Certificate
 
-		for _, chain := range verifiedChains {
-			for _, verifiedCert := range chain {
-				if verifiedCert.IsCA {
-					verifiedCerts = append(verifiedCerts, verifiedCert)
+		go func(ep string, rl *revocationList, wg *sync.WaitGroup) {
+			err := v.updateCRL(ep, rl)
+			if err != nil {
+				// Connections containing a certificate pointing to this CRL will be accepted until its current.list.NextUpdate.
```

Comment:
a CRL not being able to be updated always deserves logging I think?

Why only if `rl` (bad var name btw) is set?

Reply:
It is always logged, I just don't want to spam the logs with errors when it fails every once in a while? Maybe reduce the max failure interval to 4 hours? (CSP of all TSP claim very high uptime, so maybe it can always be logged at error level?)

Reply:
changed `rl` -> `crl`

Reply:
Now always logs on error level when it is missing and fails to download